src/Type/Php/SubstrDynamicReturnTypeExtension.php

```diff
@@ -0,0 +1,71 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Type\Php;
+
+use PhpParser\Node\Expr\FuncCall;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\FunctionReflection;
+use PHPStan\Reflection\ParametersAcceptorSelector;
+use PHPStan\Type\Accessory\AccessoryNonEmptyStringType;
+use PHPStan\Type\Constant\ConstantIntegerType;
+use PHPStan\Type\DynamicFunctionReturnTypeExtension;
+use PHPStan\Type\IntegerRangeType;
+use PHPStan\Type\IntersectionType;
+use PHPStan\Type\StringType;
+use PHPStan\Type\Type;
+
+class SubstrDynamicReturnTypeExtension implements DynamicFunctionReturnTypeExtension
+{
+
+	public function isFunctionSupported(FunctionReflection $functionReflection): bool
+	{
+		return $functionReflection->getName() === 'substr';
+	}
+
+	public function getTypeFromFunctionCall(
+		FunctionReflection $functionReflection,
+		FuncCall $functionCall,
+		Scope $scope
+	): \PHPStan\Type\Type
+	{
+		$args = $functionCall->args;
+		if (count($args) === 0) {
+			return ParametersAcceptorSelector::selectSingle($functionReflection->getVariants())->getReturnType();
+		}
+
+		if (count($args) >= 2) {
+			$string = $scope->getType($args[0]->value);
+			$offset = $scope->getType($args[1]->value);
+
+			$negativeOffset = $this->getIntValue($offset) < 0;
+			$zeroOffset = $this->getIntValue($offset) === 0;
+			$positiveLength = false;
```

Comment:
You should ask about whether a number is positive or negative this way:

```php
IntegerRangeType::fromInterval(null, -1)->isSuperTypeOf($offset)->yes() // is negative
IntegerRangeType::fromInterval(1, null)->isSuperTypeOf($offset)->yes() // is positive
```

It also accounts for unions of ConstantIntegerType, like `1|3`.

Reply:
thx, fixed.

Reply:
does this mean I should check for a zero offset with `IntegerRangeType::fromInterval(0, 0)->isSuperTypeOf($offset)->yes()` ?

Reply:
You can check for that with:

```php
(new ConstantIntegerType(0))->isSuperTypeOf($offset)->yes()
```

Reply:
> ConstantIntegerType, like `1|3`.

also added testcases to cover that